src/bin/casr-cli.rs

```diff
@@ -629,3 +660,217 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
+
+        info!("==> <{}>", filename.magenta());
+
+        // Hash each crash in cluster
+        let mut cluster_hash: HashMap<String, (Vec<String>, i32)> = HashMap::new();
+        // Hash each class in cluster
+        let mut cluster_classes: HashMap<String, i32> = HashMap::new();
+        // Hash files
+        let mut filestems: HashSet<PathBuf> = HashSet::new();
+        for report_path in fs::read_dir(cluster).unwrap().flatten() {
+            let report = report_path.path();
+            if report.is_file()
+                && report.extension().is_some()
+                && report.extension().unwrap() == "casrep"
+            {
+                // report == .*/crash.gdb.casrep
+                let mut input = report.canonicalize().unwrap().with_extension("");
+                // input == .*/crash.gdb
+                if input.extension().is_some() && input.extension().unwrap() == "gdb" {
+                    input.set_extension("");
+                }
+                // input == .*/crash
+                if !filestems.insert(input.clone()) {
+                    continue;
+                }
+
+                let mut result: Vec<String> = Vec::new();
+                let crash = if input.exists() {
+                    input.clone()
+                } else {
+                    report
+                }
+                .to_str()
+                .unwrap()
+                .to_string();
+
+                let mut report = input.to_str().unwrap().to_string();
+                report.push_str(".casrep");
+
+                let (san_desc, san_line) = if let Some((report_sum, san_desc, san_line)) =
+                    process_report(&report, "casrep")
+                {
+                    result.push(report_sum);
+                    (san_desc, san_line)
+                } else {
+                    (String::new(), String::new())
+                };
+
+                let report = report.replace(".casrep", ".gdb.casrep");
+                let (casr_gdb_desc, casr_gdb_line) =
+                    if let Some((report_sum, casr_gdb_desc, casr_gdb_line)) =
+                        process_report(&report, "gdb.casrep")
+                    {
+                        result.push(report_sum);
+                        (casr_gdb_desc, casr_gdb_line)
+                    } else {
+                        (String::new(), String::new())
+                    };
+
+                if result.is_empty() {
+                    warn!("Skipping {}. Cannot read casrep.", crash);
+                    continue;
+                } else {
+                    result.push(crash);
+                }
+
+                let mut hash = String::new();
+                hash.push_str(san_desc.as_str());
+                hash.push_str(san_line.as_str());
+                hash.push_str(casr_gdb_desc.as_str());
+                hash.push_str(casr_gdb_line.as_str());
+
+                let san_cnt = cluster_classes.get(&san_desc).unwrap_or(&0) + 1;
```

Comment:
What will be if `san_desc` is empty? We have only .gdb.casreps

Reply:
Maybe we should check this case?

Reply:
done